### ISSUE TITLE
nbdserver : notify the orchestrator about failover to slave

### ISSUE THREAD

**Issue by iwanbk** [zero-os/0-Disk#284]
Moved from #38

**Comment by iwanbk** [comment]
@zaibon please define how nbdserver should notify ays, orchestrator, or other interested parties

**Comment by zaibon** [comment]
The nbdserver need to write on stderr when it detects something is wrong. This stderr stream will is going to be monitored by AYS and self-healing action will be actived based on the information the nbdserver sends.

See https://github.com/zero-os/0-orchestrator/issues/585 for detail.

**Comment by GlenDC** [comment]
Depends upon #300.

**Comment by GlenDC** [comment]
This issue is about handling the situation where the solution as described in #357 should fail. That is to say, what we start to read/write from/to a slave server (instead of the broken primary server), and notice that the slave server is unavailable as well (due to whatever critical error). In this case we should notify the 0-orchestrator about the failure of slave server X from slave cluster Y used by vdisk Z.

**Comment by GlenDC** [comment]
So the remaining question for this issue is: What to do with the data that was supposed to be written on the slave server which is now unavailable. And the only reason that we are writing/reading to/from a slave server is because the primary server it was supposed to go to / come from failed just before. So what to do with data that has to be written to a server, while both the slave and primary server for that index are unavailable. 

Also reading from an index, for which both the slave and primary server are down, is impossible I suppose, at least for the time being. I suppose we could restore that content using tlog transactions, even though there is no real way to restore just one server. But that would be quite a slow process and I'm not sure we should do this at this stage, as it would block the program for quite a while. So perhaps a read action in this scenario should just fail?

**Comment by iwanbk** [comment]
> This issue is about handling the situation where the solution as described in #357 should fail

No, it is about when #38 (Failover to slave ardb's if the main ardb is failing) happens.

**Comment by GlenDC** [comment]
@iwanbk we're talking about the same thing... issue #357 is redo-ing a part of what was previously implemented in #38...

**Comment by iwanbk** [comment]
> @iwanbk we're talking about the same thing... issue #357 is redo-ing a part of what was previously implemented in #38...

Haa... OK :)

Because your next comment make me confused.

> So the remaining question for this issue is: What to do with the data that was supposed to be written on the slave server which is now unavailable........

**Comment by GlenDC** [comment]
Yeah I don't blame you for being confused, I'm confused around this stuff too. No worries, let's discuss this today or tomorrow together with @zaibon, and let's figure some of this stuff out. So we're all on the same page.

**Comment by iwanbk** [comment]
I think your next comment deserve separate issue, please create new issue so we don't mixing thing.
As this issue is only about notification.

**Comment by GlenDC** [comment]
Notification is in most cases one line of code. So honestly I'm not afraid for mixing things in this case... Either way, let's handle these slave-ardb related issues and such after we had a meeting and are on the same page.

**Comment by iwanbk** [comment]
> Notification is in most cases one line of code. So honestly I'm not afraid for mixing things in this case.

Because notification is simple thing, it is even worse to mix it, because people (at least me) assume simple thing in simple issue.

> Either way, let's handle these slave-ardb related issues and such after we had a meeting and are on the same page.

With creating a proper issue, people (who has concern) will not come to the meeting with blank mind.

**Comment by GlenDC** [comment]
Fair points, I'll clean up these issues tomorrow.

**Comment by iwanbk** [comment]
> Fair points, I'll clean up these issues tomorrow.

thanks @GlenDC
